[doc] Improve dev_install for Windows and LLVM 10 (#1202)
taichi-dev/taichi: docs/dev_install.rst

Note that on Linux/OS X, ``clang`` is the only supported compiler for compiling the Taichi compiler. On Windows only MSVC supported.

Installing Dependencies
-----------------------

- Make sure you are using Python 3.6/3.7/3.8
- Install Python dependencies:

  .. code-block:: bash

>     python3 -m pip install --user setuptools astpretty astor pybind11
>     python3 -m pip install --user pytest pytest-rerunfailures pytest-xdist yapf Pillow
>     python3 -m pip install --user numpy GitPython coverage colorama autograd

Reviewer: ```suggestion
    python3 -m pip install --user setuptools astpretty astor pybind11 Pillow
    python3 -m pip install --user pytest pytest-rerunfailures pytest-xdist yapf
    python3 -m pip install --user numpy GitPython coverage colorama autograd
```

- Make sure you have ``clang`` with version >= 7

  * On Windows: Download ``clang-8`` via `this link <https://releases.llvm.org/8.0.0/LLVM-8.0.0-win64.exe>`_.
    Make sure you add the ``bin`` folder containing ``clang.exe`` to the ``PATH`` environment variable.

>   * On OS X: you don't need to do anything.

Author: I see, thanks for mentioning that! Could you push to this PR about this or open a future PR on this topic?

>   * On OS X: you don't need to do anything.

Reviewer 2: Let's keep this PR related to Windows. 

Meanwhile, when testing LLVM 10 under macOS, I'll also test its compatibility with the currently working(Xcode 11.3) and the latest toolchains. Then we open a PR to add all the notes regarding installation on macOS, just like this one. But hopefully, with the updates, we can achieve the ideal "you don't need to do anything".

  * On Ubuntu, execute ``sudo apt install libtinfo-dev clang-8``.

  * On other Linux distributions, please build clang 8.0.1 from source:

    .. code-block:: bash

        wget https://github.com/llvm/llvm-project/releases/download/llvmorg-8.0.1/cfe-8.0.1.src.tar.xz
        tar xvJf cfe-8.0.1.src.tar.xz
        cd cfe-8.0.1.src
        mkdir build
        cd build
        cmake ..
        make -j 8
        sudo make install

- Make sure you have LLVM 8.0.1/10.0.0. Note that Taichi uses a customized LLVM so the pre-built binaries from the LLVM official website or other sources probably doesn't work.

>   Here we provide LLVM 8.0.1 binaries customized for Taichi, which may or may not work depending on your system environment:

Reviewer 2: Although not really related to this PR, what kind of customizations happen in the pre-built binaries? I believe this is related to tasks in #655.

Author: Basically the additional CMake options: https://github.com/taichi-dev/taichi/pull/1202/files#diff-97322f732b832df07874b286a638a2a1R66

I guess RTTI is what leads to incompatibilities (although I haven't confirmed).

  `Linux <https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1-linux-x64.zip>`_,
  `OS X <https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1.zip>`_,
  `Windows <https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1-msvc2017.zip>`_.

   If the downloaded LLVM does not work, please build from source:

>   - On Linux or OS X:

Reviewer: ```suggestion
  * On Linux or OS X:
```

>   - On Linux or OS X:

Reviewer: ```suggestion
  * On Linux or OS X:
```

      .. code-block:: bash

        wget https://github.com/llvm/llvm-project/releases/download/llvmorg-8.0.1/llvm-8.0.1.src.tar.xz
        tar xvJf llvm-8.0.1.src.tar.xz
        cd llvm-8.0.1.src
        mkdir build
        cd build
        cmake .. -DLLVM_ENABLE_RTTI:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON
        # If you are building on NVIDIA Jetson TX2, use -DLLVM_TARGETS_TO_BUILD="ARM;NVPTX"
        make -j 8
        sudo make install

>   - On Windows:

Reviewer: ```suggestion
  * On Windows:
```

    .. code-block:: bash

      # LLVM 8.0.1 + MSVC 2017
      cmake .. -G"Visual Studio 15 2017 Win64"  -DLLVM_ENABLE_RTTI:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON -Thost=x64 -DLLVM_BUILD_TESTS:BOOL=OFF -DCMAKE_INSTALL_PREFIX=installed

      # LLVM 10.0.0 + MSVC 2019
      cmake .. -G"Visual Studio 16 2019" -A x64  -DLLVM_ENABLE_RTTI:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON -Thost=x64 -DLLVM_BUILD_TESTS:BOOL=OFF -DCMAKE_INSTALL_PREFIX=installed

>     Then open ``LLVM.sln`` and use Visual Studio 2017+ to build. Please make sure you are using the ``Release`` configuration. After building the ``INSTALL`` project (under folder ``CMakePredefinedTargets`` in the Solution Explorer window).
>     After the build is complete, find your LLVM binaries and headers in ``build/installed``.
>     If you use MSVC 2019, **make sure you use C++17** for the ``INSTALL`` project.

Reviewer: ```suggestion
    1. Then open and build ``LLVM.sln`` with Visual Studio 2017+. Make sure you are using the ``Release`` configuration.
    2. Then build the ``INSTALL`` project under folder ``CMakePredefinedTargets`` in the Solution Explorer window.
    
    .. note::
    
        After the build is complete, you may find your LLVM binaries and headers in ``build/installed``.
        If you use MSVC 2019, **make sure you use C++17** for the ``INSTALL`` project.
```

Author: Thanks! I'll use a bulleted list, but the `note` does not quite necessary here.

>     Please add ``build/installed/bin`` to ``PATH``.
>     Later, when you build Taichi using ``CMake``, set ``LLVM_DIR`` to ``build/installed/lib/cmake/llvm``.

Reviewer: ```suggestion
    Then configure the environment variables:
     1. Add ``build/installed/bin`` to ``%PATH%``.
     2. Set ``%LLVM_DIR%`` to ``build/installed/lib/cmake/llvm``.
```

> - On Windows, if you use the pre-built LLVM for Taichi, please add ``$LLVM_FOLDER/bin`` to ``PATH``.
>   Later, when you build Taichi using ``CMake``, set ``LLVM_DIR`` to ``$LLVM_FOLDER/lib/cmake/llvm``.

Reviewer: ```suggestion
- On Windows, if you use the pre-built LLVM for Taichi, please:
   1. Add ``<llvm folder>/bin`` to ``%PATH%``.
   2. Set ``%LLVM_DIR%`` to ``<llvm folder>/lib/cmake/llvm``.
```

Author: Actually, the `LLVM_DIR` is a CMake option instead of an environment variable.
